binderhub/repoproviders.py

```diff
@@ -351,6 +351,58 @@ def get_build_slug(self):
         return "dataverse-" + escapism.escape(self.identifier, escape_char="-").lower()
 
 
+class HydroshareProvider(RepoProvider):
+    """Provide contents of a Hydroshare resource
+    Users must provide a spec consisting of the Hydroshare resource id.
+    """
+    name = Unicode("Hydroshare")
+    url_regex = re.compile(r".*([0-9a-f]{32}).*")
+
+    def _parse_resource_id(self, spec):
+        match = self.url_regex.match(spec)
+        if not match:
+            raise ValueError("The specified Hydroshare resource id was not recognized.")
+        resource_id = match.groups()[0]
+        return resource_id
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        client = AsyncHTTPClient()
+        self.resource_id = self._parse_resource_id(self.spec)
+        req = HTTPRequest("https://www.hydroshare.org/hsapi/resource/{}/scimeta/elements".format(self.resource_id),
+                          user_agent="BinderHub")
+        r = yield client.fetch(req)
+        def parse_date(json_body):
+            json_response = json.loads(json_body)
+            date = next(
+                item for item in json_response["dates"] if item["type"] == "modified"
+            )["start_date"]
+            # Hydroshare timestamp always returns the same timezone, so strip it
+            date = date.split(".")[0]
+            parsed_date = datetime.strptime(date, "%Y-%m-%dT%H:%M:%S")
+            epoch = parsed_date.replace(tzinfo=timezone(timedelta(0))).timestamp()
+            # truncate the timestamp
+            return str(int(epoch))
+        # date last updated is only good for the day... probably need something finer eventually
+        self.record_id = "{}.v{}".format(self.resource_id, parse_date(r.body))
+        return self.record_id
+
+    async def get_resolved_spec(self):
+        # Hydroshare does not provide a history, resolves to repo url
+        return self.get_repo_url()
+
+    async def get_resolved_ref_url(self):
+        # Hydroshare does not provide a history, resolves to repo url
+        return self.get_repo_url()
```

Comment:
get_resolved_spec and get_resolved_ref_url were two methods that I've implemented since the original review.  With what I gleaned from the documentation, resolving to get_repo_url() made sense for HydroShare.  If this is incorrect, I'll need suggestions to guide a different implementation.